src/braket/tracking/tracker.py

```diff
@@ -0,0 +1,276 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+from decimal import Decimal
+from functools import lru_cache, singledispatch
+from json import loads
+from typing import Any, Dict, List
+
+import braket.aws
+from braket.tracking.tracking_context import deregister_tracker, register_tracker
+from braket.tracking.tracking_events import _TaskCompletionEvent, _TaskCreationEvent, _TaskGetEvent
+
+MIN_SIMULATOR_DURATION = 3000
+
+
+class Tracker:
+    """
+    Amazon Braket cost tracker.
+    Use this class to track costs incurred from quantum tasks on Amazon Braket.
+    """
+
+    def __init__(self):
+        self._resources = {}  # Key = quantum_task_arn
+
+    def __enter__(self):
+        register_tracker(self)
+        return self
+
+    def __exit__(self, *args):
+        deregister_tracker(self)
+
+    def start(self):
+        """
+        Start tracking resources with this tracker.
+        """
+        return self.__enter__()
+
+    def stop(self):
+        """
+        Stop tracking resources with this tracker.
+        """
+        return self.__exit__()
+
+    def receive_event(self, event):
+        _recieve_internal(event, self._resources)
+
+    def tracked_resources(self) -> List[str]:
+        """
+        Resources tracked by this tracker.
+
+        Returns:
+            List[str]: The list of task ids for tasks tracked by this tracker.
+        """
+        return list(self._resources.keys())
+
+    def qpu_tasks_cost(self) -> Decimal:
+        """
+        Estimate cost of all quantum tasks tracked by this tracker that use Braket qpu devices.
+
+        Returns:
+            Decimal: The estimated total cost in USD
+        """
+        total_cost = Decimal(0)
+        for task_arn, details in self._resources.items():
+            if "qpu" in details["device"]:
+                total_cost = total_cost + _get_qpu_task_cost(task_arn, details)
+        return total_cost
+
+    def simulator_tasks_cost(self) -> Decimal:
+        """
+        Estimate cost of all quantum tasks tracked by this tracker using Braket simulator devices.
+
+        Returns:
+            Decimal: The estimated total cost in USD
+        """
+        total_cost = Decimal(0)
+        for task_arn, details in self._resources.items():
+            if "qpu" not in details["device"]:
```

Comment:
I think checking for `"simulator" in details["device"]` might help protect against the case where (if) a third kind of device is introduced, it won't be mistaken for a simulator.

Reply:
Done